cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,183 @@
+	for i := 0; i < timeout; i++ {
+		var res []interface{}
+		res, err := conn.Eval(
+			fmt.Sprintf("return box.info.vclock[%d]", masterUpgradeInfo.IID),
+			[]any{}, connector.RequestOpts{})
+		if err != nil || len(res) == 0 {
+			return err
+		}
+		var lsn uint32
+		switch v := res[0].(type) {
+		case uint16:
+			lsn = uint32(v)
+		case uint32:
+			lsn = v
+		}
+		if lsn >= masterUpgradeInfo.LSN {
+			return nil
+		}
+		time.Sleep(time.Second)
+	}
+	return errors.New("")
+}
+
+func GetAllowedReplicasets(allReplicasets replicaset.Replicasets,
+	pendingReplicasetAliases []string) ([]replicaset.Replicaset, error) {
+	if len(pendingReplicasetAliases) == 0 {
+		return allReplicasets.Replicasets, nil
+	}
+
+	replicasetMap := make(map[string]replicaset.Replicaset)
+	for _, rs := range allReplicasets.Replicasets {
+		replicasetMap[rs.Alias] = rs
+	}
+
+	var allowedReplicasets []replicaset.Replicaset
+
+	for _, alias := range pendingReplicasetAliases {
+		replicaset, exists := replicasetMap[alias]
+		if !exists {
+			return nil, fmt.Errorf("replicaset with alias %q doesn't exist", alias)
+		}
+		allowedReplicasets = append(allowedReplicasets, replicaset)
+	}
+
+	return allowedReplicasets, nil
+}
+
+func Upgrade(replicasets replicaset.Replicasets, pendingReplicasetAliases []string,
+	lsnTimeout int) error {
+	allowedReplicasets, err := GetAllowedReplicasets(replicasets, pendingReplicasetAliases)
+	if err != nil {
+		return err
+	}
+
+	printReplicasetStatus := func(alias, status string) {
+		fmt.Printf("• %s: %s\n", alias, status)
+	}
+
+	for _, rs := range allowedReplicasets {
+		var masterRun *running.InstanceCtx = nil
+		var replicRun []running.InstanceCtx
+
+		for _, inst := range rs.Instances {
+			run := inst.InstanceCtx
+			fullInstanceName := running.GetAppInstanceName(run)
+			conn, err := connector.Connect(connector.ConnectOpts{
+				Network: "unix",
+				Address: run.ConsoleSocket,
+			})
+			if err != nil {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+
+			res, err := conn.Eval(
+				"return (type(box.cfg) == 'function') or box.info.ro",
+				[]any{}, connector.RequestOpts{})
+			if err != nil || len(res) == 0 {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+
+			if !res[0].(bool) {
+				if masterRun != nil {
+					printReplicasetStatus(rs.Alias, "error")
+					return fmt.Errorf("[%s]: %s and %s are both masters",
+						rs.Alias, running.GetAppInstanceName(*masterRun),
+						fullInstanceName)
+				}
+				masterRun = &run
+			} else {
+				replicRun = append(replicRun, run)
+			}
+		}
+		if masterRun == nil {
+			printReplicasetStatus(rs.Alias, "error")
+			return fmt.Errorf("[%s]: has not master instance", rs.Alias)
+		}
+		var conn connector.Connector
+		conn, err = connector.Connect(connector.ConnectOpts{
+			Network: "unix",
+			Address: masterRun.ConsoleSocket,
+		})
```

Comment:
Can we re-use a connection from the loop above? It seems like we don't have to connect twice.

The same for the replica connections below.

Reply:
Subject to discussion. See my [comment](https://github.com/tarantool/tt/pull/936#issuecomment-2393858831) below.

Reply:
I’m now reusing the connection, but I believe we should implement the `conn.isAlive` method to verify if the connection is active, especially for non-unix socket connections.

Reply:
AFAIU, this `conn.isAlive` check wouldn't affect our scenario much.

The pattern of accesses to the instances is a kind of RPC (remote procedure call), not PubSub (publish-subscribe). So, if we're going to check this `conn.isAlive` property before an eval, it is the same as just attempt to perform the eval itself.

Moreover, an instance may change its RO/RW status or go offline while our eval request goes over the wire or even while it is executed on the instance. A client side `conn.isAlive` property doesn't guarantee that the next request is executed successfully. It means that we anyway should handle eval's failure, disregarding whether we check `conn.isAlive` or not.

*(My question is resolved. Feel free to close this discussion.)*